File: QytetetRuby/lib/controlador_qytetet.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "qytetet"
require_relative "vista_textual_qytetet"
require_relative "jugador"
require_relative "sorpresa"
require_relative "tipo_sorpresa"
require_relative "tablero"
require_relative "titulo_propiedad"
require_relative "tipo_casilla"
require_relative "casilla"
require_relative "metodo_salir_carcel"
require "singleton"
require_relative "dado"
require_relative "especulador"
require_relative "calle"

module InterfazTextualQytetet
 
  class ControladorQytetet
    attr_reader :vista
    def initialize
      @juego
      @jugador
      @casilla
      @vista = VistaTextualQytetet.new
    end
    
    def inicializacion_juego
      @juego = ModeloQytetet::Qytetet.instance
      nombres = @vista.obtener_nombre_jugadores 
      @juego.inicializar_juego(nombres)
      @jugador = @juego.jugador_actual
      @casilla = @jugador.casilla_actual
      
      puts @juego
      puts "=======" + "jugador Actual" + "========"
      puts @jugador
      gets.chomp
    end    
    
    def desarrollo_juego
      
      
      loop do
        bancarrota = @jugador.saldo <= 0
        if (@jugador.encarcelado)
          metodo = @vista.menu_salir_carcel
          if (metodo == 0)
            libre = @juego.intentar_salir_carcel(ModeloQytetet::Metodo_salir_carcel::TIRANDODADO)
           
          elsif (metodo == 1)
            libre = @juego.intentar_salir_carcel(ModeloQytetet::Metodo_salir_carcel::PAGANDOLIBERTAD)
          end
        end
        if (libre || !@jugador.encarcelado)
          no_tiene_propietario = @juego.jugar
          bancarrota = @jugador.saldo <= 0
          if (!bancarrota)
            if (!@jugador.encarcelado)            
              if (@jugador.casilla_actual.tipo == ModeloQytetet::TipoCasilla::SORPRESA) #preguntar SI CON JUGADOR O CON CASILLA.TIPO
                no_tiene_propietario = @juego.aplicar_sorpresa
               # @jugador.casilla_actual = 
                if (!@jugador.encarcelado)
                  bancarrota = @jugador.saldo <= 0
                  if(!bancarrota)
                    if (@jugador.casilla_actual.tipo == ModeloQytetet::TipoCasilla::CALLE)
                      if (!@jugador.casilla_actual.tengo_propietario)
                        eleccion = @vista.elegir_quiero_comprar
                        if (eleccion == 0)
                          @juego.comprar_titulo_de_propiedad
                        end
                      end
                    end
                  end
                end
              elsif (@jugador.casilla_actual.tipo == ModeloQytetet::TipoCasilla::CALLE)
                if(!@jugador.casilla_actual.tengo_propietario)
                  eleccion = @vista.elegir_quiero_comprar
                  if (eleccion == 0)
                    @juego.comprar_titulo_de_propiedad
                  end
                end
              end
            end
          end
        end
        bancarrota = @jugador.saldo <= 0
        if (!@jugador.encarcelado && !bancarrota && @jugador.tengo_propiedad)
          opcion = @vista.menu_gestion_inmobiliaria
          while opcion != 0 
            titulo = elegir_propiedad(@jugador.propiedades)
            
            if (opcion == 1)
              puedo_edificar = @juego.edificar_casa(titulo.casilla)
              if (!puedo_edificar)
                @vista.mostrar("No se pueden poner mas casas")
              end
               #el parametro que se le pasa es casilla actual del jugador?
            elsif (opcion == 2)
              puedo_edificar = @juego.edificar_hotel(titulo.casilla)
              if (!puedo_edificar)
                @vista.mostrar("No se pueden poner mas hoteles")
              end
            elsif (opcion == 3)
              puedo_vender = @juego.vender_propiedad(titulo.casilla)
              if (!puedo_vender)
                @vista.mostrar("No se puede vender")
              end
            elsif (opcion == 4)
              hipotecar = @juego.hipotecar_propiedad(titulo.casilla)
              if (!hipotecar)
                @vista.mostrar("No se puede hipotecar")
              end
            elsif (opcion == 5)
              cancelar = @juego.cancelar_hipoteca(titulo.casilla)
              if (!cancelar)
                @vista.mostrar("No se puede cancelar hipoteca")
              end
            end
            opcion = @vista.menu_gestion_inmobiliaria
          
            end
        end
           
        
        bancarrota = @jugador.saldo <= 0
        if (!bancarrota)
          @jugador = @juego.siguiente_jugador
          puts "Ha cambiado de jugador"
          puts "===" + " Jugador Actual " + "===="
          puts @jugador
        end
       
        break if bancarrota
      end
      puts "UN JUGADOR HA ENTRADO EN BANCARROTA"
      puts @juego.obtener_ranking
      
    end
    
    
    def elegir_propiedad(propiedades) # lista de propiedades a elegir
        vista.mostrar("\tCasilla\tTitulo")
        
        listaPropiedades= Array.new
        for prop in propiedades  # crea una lista de strings con numeros y nombres de propiedades
                propString= prop.casilla.numero_casilla.to_s+' '+prop.nombre 
                listaPropiedades<<propString
        end
        seleccion = vista.menu_elegir_propiedad(listaPropiedades)  # elige de esa lista del menu
        propiedades.at(seleccion)
   end
   def self.main
     controlador = ControladorQytetet.new
     controlador.inicializacion_juego
     controlador.desarrollo_juego

     
   end
   ControladorQytetet.main
  end
  
end

File: QytetetJava/src/InterfazTextualQytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package InterfazTextualQytetet;

import java.util.ArrayList;
import modeloqytetet.Qytetet;
import modeloqytetet.Jugador;
import modeloqytetet.Casilla;
import modeloqytetet.Calle;
import modeloqytetet.OtraCasilla;
import modeloqytetet.Dado;
import modeloqytetet.MetodoSalirCarcel;
import modeloqytetet.Tablero;
import modeloqytetet.TipoCasilla;
import modeloqytetet.TituloPropiedad;
/**
 *
 * @author juliu
 */
public class ControladorQytetet {
    public static Qytetet juego;
    public static Jugador jugador;
    public static Casilla casilla;
    public static VistaTextualQytetet vista = new VistaTextualQytetet();
    
    public void inicializacionJuego(){
        juego = Qytetet.getInstance();
        ArrayList<String> nombre = new ArrayList<>();
        nombre = vista.obtenerNombreJugadores();
       
        juego.inicializarJuego(nombre);
        jugador = juego.getJugadorActual(); 
        casilla = jugador.getCasillaActual();
        
        juego.toString();
        System.out.println("====" + "Jugador Actual" + "====");
        System.out.println(jugador);
        System.out.println("\n");
        
    }
    
    public void desarrolloJuego(){
        boolean bancarrota;
        bancarrota = jugador.getSaldo() <= 0;
        do{
            boolean libre = false;
            if (jugador.getEncarcelado()){
                int metodo = vista.menuSalirCarcel();
                if (metodo == 0){
                    libre = juego.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO);
                }else{
                    libre = juego.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD);
                }    
            }
            if (libre || !jugador.getEncarcelado()){
                boolean noTienePropietario = juego.jugar();
                bancarrota = jugador.getSaldo() <= 0;
                if (!bancarrota){
                    if (!jugador.getEncarcelado()){
                        if (((OtraCasilla)jugador.getCasillaActual()).getTipo() == TipoCasilla.SORPRESA){
                            noTienePropietario = juego.aplicarSorpresa();
                            if (!jugador.getEncarcelado()){
                                bancarrota = jugador.getSaldo() <= 0;
                                if (!bancarrota){
                                    if (((OtraCasilla)jugador.getCasillaActual()).getTipo() == TipoCasilla.CALLE){
                                        if (!((Calle)jugador.getCasillaActual()).tengoPropietario()){
                                            int eleccion = vista.elegirQuieroComprar();
                                            if (eleccion == 0){
                                                juego.comprarTituloPropiedad();
                                            }
                                        }
                                    }
                                }
                            }
                        }else if (((OtraCasilla)jugador.getCasillaActual()).getTipo() == TipoCasilla.CALLE){
                            if (!((Calle)jugador.getCasillaActual()).tengoPropietario()){
                                int eleccionDos = vista.elegirQuieroComprar();
                                if(eleccionDos == 0){
                                    juego.comprarTituloPropiedad();
                                }
                            }
                        }
                    }
                }
            }
           
            bancarrota = jugador.getSaldo() <= 0;
            if (!jugador.getEncarcelado() && !bancarrota && jugador.tengoPropiedad()){
                int opcion = vista.menuGestionInmobiliaria();
                while (opcion != 0){
                    TituloPropiedad titulo = elegirPropiedad(jugador.getPropiedades());
                    if (opcion == 1){
                        boolean puedoEdificar = juego.edificarCasa(titulo.getCasilla());
                        if (!puedoEdificar){
                            vista.mostrar("No puedes edificar mas casas");
                        }
                    }else if (opcion == 2){
                        boolean puedoEdificar = juego.edificarHotel(titulo.getCasilla());
                        if (!puedoEdificar){
                            vista.mostrar("No puedes edificar mas hoteles");
                        }
                    }else if (opcion == 3){
                        boolean puedoVender = juego.venderPropiedad(titulo.getCasilla());
                        if (!puedoVender){
                            vista.mostrar("No puedes vender");
                        }
                    }else if (opcion == 4){
                        boolean puedoHipotecar = juego.hipotecarPropiedad(titulo.getCasilla());
                        if (!puedoHipotecar){
                            vista.mostrar("No puedes hipotecar la propiedad");
                        }
                    }else if (opcion == 5){
                        juego.cancelarHipoteca(titulo.getCasilla());
                    }
                opcion = vista.menuGestionInmobiliaria();
                }
            }
            
        bancarrota = jugador.getSaldo() <= 0;
        if (!bancarrota){
            jugador = juego.siguienteJugador();
            System.out.println("Ha cambiado de jugador \n");
            System.out.println("====" + "Jugador Actual" + "====");
            System.out.println(jugador);
            System.out.println("\n");
        }
        }while (!bancarrota);
        System.out.println("JUEGO FINALIZADO, UN JUGADOR HA ENTRADO EN BANCARROTA");
        juego.obtenerRanking().toString();
    }
    
    public TituloPropiedad elegirPropiedad(ArrayList<TituloPropiedad> propiedades){ 
 //este metodo toma una lista de propiedades y genera una lista de strings, con el numero y nombre de las propiedades
 //luego llama a la vista para que el usuario pueda elegir.
        vista.mostrar("\tCasilla\tTitulo");
        int seleccion = 0;
        ArrayList<String> listaPropiedades= new ArrayList();
        for ( TituloPropiedad prop: propiedades) {
                listaPropiedades.add( "\t"+prop.getCasilla().getNumeroCasilla()+"\t"+prop.getNombre());
        seleccion=vista.menuElegirPropiedad(listaPropiedades);  
        
         }
        return propiedades.get(seleccion);
    }
 
    public static void main(String[] args){
        ControladorQytetet controlador = new ControladorQytetet();
        controlador.inicializacionJuego();
        controlador.desarrolloJuego();
    }
    
    
    
}
